From e14c0832fe2a82015fd837cfa512517734f9f05c Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Tue, 6 Nov 2018 15:10:04 +0100
Subject: [PATCH 2/2] [SQUSH][DNM] Merge tag 'android-9.0.0_r16' into
 lineage-16.0

commit c032802dca743fb31957e8702682d920c1d73f8f
Merge: 62f1871b 2a4f688b
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Tue Aug 21 03:12:45 2018 +0000

    Snap for 4966094 from 2a4f688b3ae3b4151ab4aab8dfad81e8c2a2e7e4 to pi-qpr1-release

    Change-Id: Ifd42b52264567ae6370fcc18ded5d073c0b2baa1

commit 2a4f688b3ae3b4151ab4aab8dfad81e8c2a2e7e4
Author: George Chang <georgekgchang@google.com>
Date:   Mon Jul 9 19:06:08 2018 +0800

    Skip setEnableHostRouting when switching to screen off

    Bug: 111246269
    Test: Enable/Disable, Check HCE status at screen on/off
    Change-Id: I8f52e7fa33db38e3ed67f9b0f84a3836c1199b22
    (cherry picked from commit 926547b6118ebd4ccff5489b6df1abedb9a1502a)

commit 62f1871bc3ab5a854922daea2e53c70940aa18c4
Merge: a2ea9593 a9ef080f
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Tue Jul 31 03:10:43 2018 +0000

    Snap for 4922203 from a9ef080f9964f46d4983bf7637a2c52a6e0eadd2 to pi-qpr1-release

    Change-Id: I7618716c7271dcba29c2254d536fe2997e586c14

commit a9ef080f9964f46d4983bf7637a2c52a6e0eadd2
Author: Yoshiaki Naka <yoshiaki.naka@sony.com>
Date:   Fri Jun 8 10:05:34 2018 +0900

    SecureElementService may be unavailable when NfcService is started

    Currently NfcService attempts to get an interface to
    SecureElementService only in its constructor, so there is no chance to
    do it later again if it fails. Introducing a lazy initialization
    mechanism is required.

    Bug: 111775064
    Bug: 110121800
    Test: Confirmed that NfcService gets SEService even if it failed once.

    Change-Id: I2b3c873d7661cde63af8c9f8c05945c3568f012f

Change-Id: Ib462a63265910cb125f01582659c874f8802cb17
---
 src/com/android/nfc/NfcService.java | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)
 mode change 100755 => 100644 src/com/android/nfc/NfcService.java

diff --git a/src/com/android/nfc/NfcService.java b/src/com/android/nfc/NfcService.java
old mode 100755
new mode 100644
index 0f12ac78..41e7e707
--- a/src/com/android/nfc/NfcService.java
+++ b/src/com/android/nfc/NfcService.java
@@ -518,6 +518,14 @@ public class NfcService implements DeviceHostListener {
                 Context.SECURE_ELEMENT_SERVICE));
     }
 
+    private boolean isSEServiceAvailable() {
+        if (mSEService == null) {
+            mSEService = ISecureElementService.Stub.asInterface(ServiceManager.getService(
+                    Context.SECURE_ELEMENT_SERVICE));
+        }
+        return (mSEService != null);
+    }
+
     void initSoundPool() {
         synchronized (this) {
             if (mSoundPool == null) {
@@ -1779,10 +1787,9 @@ public class NfcService implements DeviceHostListener {
             paramsBuilder.setTechMask(techMask);
             paramsBuilder.setEnableLowPowerDiscovery(false);
             paramsBuilder.setEnableP2p(false);
-        } else {
         }
 
-        if (mIsHceCapable && mReaderModeParams == null) {
+        if (mIsHceCapable && mScreenState >= ScreenStateHelper.SCREEN_STATE_ON_LOCKED && mReaderModeParams == null) {
             // Host routing is always enabled at lock screen or later, provided we aren't in reader mode
             paramsBuilder.setEnableHostRouting(true);
         }
@@ -2261,7 +2268,7 @@ public class NfcService implements DeviceHostListener {
         }
 
         private void sendOffHostTransactionEvent(byte[] aid, byte[] data, byte[] readerByteArray) {
-            if (mSEService == null || mNfcEventInstalledPackages.isEmpty()) {
+            if (!isSEServiceAvailable() || mNfcEventInstalledPackages.isEmpty()) {
                 return;
             }
 
@@ -2301,7 +2308,7 @@ public class NfcService implements DeviceHostListener {
 
         /* Returns the list of packages that have access to NFC Events on any SE */
         private ArrayList<String> getSEAccessAllowedPackages() {
-            if (mSEService == null || mNfcEventInstalledPackages.isEmpty()) {
+            if (!isSEServiceAvailable() || mNfcEventInstalledPackages.isEmpty()) {
                 return null;
             }
             String[] readers = null;
-- 
2.17.1

